crates/next-api/src/client_references.rs

```diff
@@ -24,50 +32,179 @@ pub enum ClientReferenceMapType {
+}
+
+impl ClientReferencesSet {
+    fn new(client_references: ClientReferencesSet) -> Vc<Self> {
+        Self::cell(client_references)
+    }
+
+    /// Returns all the server components that depend on the given client reference
+    pub fn server_components_for_client_reference(
+        &self,
+        module: ResolvedVc<Box<dyn Module>>,
+    ) -> impl Iterator<Item = ResolvedVc<NextServerComponentModule>> {
+        let bitmap = &self
+            .server_components_for_client_references
+            .get(&module)
+            .expect("Module should be a client reference module")
+            .0;
+
+        bitmap
+            .iter()
+            .map(|index| *self.server_components.get_index(index as usize).unwrap())
+    }
+}
 
 #[turbo_tasks::function]
 pub async fn map_client_references(
     graph: Vc<SingleModuleGraph>,
 ) -> Result<Vc<ClientReferencesSet>> {
-    let client_references = graph
-        .await?
-        .iter_nodes()
-        .map(|node| async move {
-            let module = node.module;
+    let span = tracing::info_span!("mapping client references");
+    async move {
+        let graph = graph.await?;
+        let client_references = graph
+            .iter_nodes()
+            .map(|node| async move {
+                let module = node.module;
+
+                if let Some(client_reference_module) =
+                    ResolvedVc::try_downcast_type::<EcmascriptClientReferenceModule>(module)
+                {
+                    Ok(Some((
+                        module,
+                        ClientReferenceMapType::EcmascriptClientReference {
+                            module: client_reference_module,
+                            ssr_module: ResolvedVc::upcast(
+                                client_reference_module.await?.ssr_module,
+                            ),
+                        },
+                    )))
+                } else if let Some(client_reference_module) =
+                    ResolvedVc::try_downcast_type::<CssClientReferenceModule>(module)
+                {
+                    Ok(Some((
+                        module,
+                        ClientReferenceMapType::CssClientReference(
+                            client_reference_module.await?.client_module,
+                        ),
+                    )))
+                } else if let Some(server_component) =
+                    ResolvedVc::try_downcast_type::<NextServerComponentModule>(module)
+                {
+                    Ok(Some((
+                        module,
+                        ClientReferenceMapType::ServerComponent(server_component),
+                    )))
+                } else {
+                    Ok(None)
+                }
+            })
+            .try_flat_join()
+            .await?
+            .into_iter()
+            .collect::<FxHashMap<_, _>>();
+
+        let mut server_components = FxIndexSet::default();
+        let mut parent_modules = FxHashMap::default();
```

Comment:
The name is a bit confusing. Maybe name it `map_module_to_server_components_bits`